### PR TITLE
Distributed K-Means on GPU

### PR DESCRIPTION
**Pull request by rlnx**
- [x] Modify K-Means algorithm to run in SPMD mode
- [x] Extend `ndarray` with `.split(...)` method
- [x] Add functions for `unique_ptr` allocated on host
- [x] Refactor tests for K-Means kernels
- [x] Add integration tests for SPMD K-Means based on thread communicator
- [x] Rework base algorithm fixtures to reuse them for distributed tests

### CODE REVIEW

File: cpp/oneapi/dal/backend/dispatcher.hpp [hunk 1]
```diff
@@ -233,7 +249,7 @@ struct kernel_dispatcher<kernel_spec<single_node_cpu_kernel, CpuKernel>,
     auto operator()(const detail::spmd_data_parallel_policy& policy, Args&&... args) const {
         return dispatch_by_device(
             policy.get_local(),
-            [&]() {
+            [&]() -> gpu_kernel_return_t<GpuKernel, Args...> {
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/backend/dispatcher.hpp:252]: Question: why the same return type is not used in operator above?

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/cluster_updater.hpp [hunk 2]
```diff
@@ -26,150 +28,153 @@ namespace bk = dal::backend;
...
+        centroids_event = kernels_fp_t::merge_reduce_centroids( //
+            queue_,
+            counters_,
+            partial_centroids_,
+            part_count_,
+            centroids,
+            { count_event, centroids_event });
+
+        auto centroids_reduce_request =
+            comm_.allreduce(centroids.flatten(queue_, { centroids_event }));
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/cluster_updater.hpp:137]: Yes, it's a bug

[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/cluster_updater.hpp:137]: Discussed offline

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                                             host_all_candidates.flatten());
+
+    // Allgather distances
+    const auto host_distances = distances.to_host(queue);
+    auto host_all_distances = pr::ndarray<Float, 2>::empty({ comm.get_rank_count(), //
+                                                             candidate_count });
+    auto distances_request = comm.allgather(host_distances.flatten(), //
+                                            host_all_distances.flatten());
+
+    auto host_all_indices = bk::make_unique_host<std::int32_t>(all_candidate_count);
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp:239]: Discussed offline

**Pull request merged**